File: backend/routes/llamadasVentasLineas.js
const express = require('express');
const router = express.Router();

const { getDb } = require('../config/db');
const { protect, authorize } = require('../middleware/auth');

let __indexesReady = false;
async function getCollection() {
  const db = getDb();
  if (!db) {
    throw new Error('DB no inicializada');
  }
  const coll = db.collection('LlamadasVentasLineas');
  if (!__indexesReady) {
    try {
      await coll.createIndex({ fecha: 1 }, { unique: true, name: 'uniq_fecha_lineas' });
      __indexesReady = true;
    } catch (_) {
      // ignorar errores de índices existentes
    }
  }
  return coll;
}

// GET /api/llamadas-ventas-lineas/:fecha -> { ok, data }
router.get('/:fecha', protect, authorize('admin','Administrador','administrador','backoffice','Backoffice'), async (req, res) => {
  try {
    const fecha = req.params.fecha;
    if (!fecha) {
      return res.status(400).json({ ok: false, message: 'Fecha requerida' });
    }
    
    const coll = await getCollection();
    const doc = await coll.findOne({ fecha });

    if (doc) {
      return res.json({ ok: true, data: doc });
    } else {
      return res.json({ ok: true, data: null });
    }
  } catch (e) {
    console.error('[LLAMADAS-VENTAS-LINEAS] GET error:', e);
    res.status(500).json({ ok: false, message: 'Error interno' });
  }
});

// GET /api/llamadas-ventas-lineas/mes/:anio/:mes -> { ok, data: [] }
router.get('/mes/:anio/:mes', protect, authorize('admin','Administrador','administrador','backoffice','Backoffice'), async (req, res) => {
  try {
    const anio = Number(req.params.anio);
    const mes = Number(req.params.mes);
    
    if (!anio || !mes || mes < 1 || mes > 12) {
      return res.status(400).json({ ok: false, message: 'Parámetros inválidos' });
    }
    
    const coll = await getCollection();
    // Buscar fechas que coincidan con el patrón YYYY-MM-*
    const regex = new RegExp(`^${anio}-${String(mes).padStart(2, '0')}-`);
    const docs = await coll.find({ fecha: { $regex: regex } }).sort({ fecha: 1 }).toArray();

    return res.json({ ok: true, data: docs });
  } catch (e) {
    console.error('[LLAMADAS-VENTAS-LINEAS] GET mes error:', e);
    res.status(500).json({ ok: false, message: 'Error interno' });
  }
});

// POST /api/llamadas-ventas-lineas -> body { fecha, equipos }
router.post('/', protect, authorize('admin','Administrador','administrador','backoffice','Backoffice'), async (req, res) => {
  try {
    const { fecha, equipos } = req.body || {};
    
    if (!fecha) {
      return res.status(400).json({ ok: false, message: 'Fecha requerida' });
    }
    
    const coll = await getCollection();
    const now = new Date();
    const username = req.user?.username || null;

    const result = await coll.updateOne(
      { fecha },
      {
        $set: {
          equipos: equipos || {},
          updatedAt: now,
          updatedBy: username,
        },
        $setOnInsert: {
          createdAt: now,
          createdBy: username,
        }
      },
      { upsert: true }
    );

    return res.json({ ok: true, upserted: !!result.upsertedId, modifiedCount: result.modifiedCount });
  } catch (e) {
    if (e && e.code === 11000) {
      return res.status(409).json({ ok: false, message: 'Conflicto de duplicado para la fecha' });
    }
    console.error('[LLAMADAS-VENTAS-LINEAS] POST error:', e);
    res.status(500).json({ ok: false, message: 'Error interno' });
  }
});

module.exports = router;

File: CRM_PYTHON/routers/llamadas_ventas_lineas.py
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from database import get_db
from deps import require_roles
from datetime import datetime
from typing import Optional, Any
import re

router = APIRouter(prefix="/api/llamadas-ventas-lineas", tags=["Llamadas Ventas Lineas"])

_ADMIN_BO_ROLES = ("admin", "Administrador", "administrador", "backoffice", "Backoffice")


class LlamadasBody(BaseModel):
    fecha: str
    equipos: Optional[Any] = None


@router.get("/mes/{anio}/{mes}")
async def llamadas_mes(anio: int, mes: int, user: dict = Depends(require_roles(*_ADMIN_BO_ROLES))):
    if mes < 1 or mes > 12:
        raise HTTPException(400, "Parámetros inválidos")
    db = get_db()
    prefix = f"{anio}-{str(mes).zfill(2)}-"
    docs = await db["LlamadasVentasLineas"].find(
        {"fecha": {"$regex": f"^{re.escape(prefix)}"}}
    ).sort("fecha", 1).to_list(None)
    for d in docs:
        d["_id"] = str(d["_id"])
    return {"ok": True, "data": docs}


@router.get("/{fecha}")
async def llamadas_por_fecha(fecha: str, user: dict = Depends(require_roles(*_ADMIN_BO_ROLES))):
    db = get_db()
    doc = await db["LlamadasVentasLineas"].find_one({"fecha": fecha})
    if doc:
        doc["_id"] = str(doc["_id"])
    return {"ok": True, "data": doc}


@router.post("/")
async def llamadas_save(body: LlamadasBody, user: dict = Depends(require_roles(*_ADMIN_BO_ROLES))):
    if not body.fecha:
        raise HTTPException(400, "Fecha requerida")
    db = get_db()
    now = datetime.utcnow()
    username = user.get("username")
    try:
        result = await db["LlamadasVentasLineas"].update_one(
            {"fecha": body.fecha},
            {
                "$set": {"equipos": body.equipos or {}, "updatedAt": now, "updatedBy": username},
                "$setOnInsert": {"createdAt": now, "createdBy": username},
            },
            upsert=True,
        )
        return {"ok": True, "upserted": result.upserted_id is not None, "modifiedCount": result.modified_count}
    except Exception as e:
        if "11000" in str(e) or "duplicate" in str(e).lower():
            raise HTTPException(409, "Conflicto de duplicado para la fecha")
        raise HTTPException(500, "Error interno")
